libraries/AP_Common/Location.cpp

```diff
@@ -223,6 +223,39 @@ float Location::get_distance(const struct Location &loc2) const
     return norm(dlat, dlng) * LOCATION_SCALING_FACTOR;
 }
 
+// return ETA (Estimated Time of Arrival) from Location to loc2 in Seconds taking current Wind & Airspeed estimates into account
+// Should be used with caution - the return value's sliding average of a reasonably long time period (ie 5..30 seconds) is suggested for estimation
+// Negative return value should never be considered as part of any sliding average, as -1 is used to indicate infinately long ETA
+// Note that AP::ahrs().wind_estimate() only appears to provide realistic estimate after flying a course including at least 180 degree turns - currently
+// Note that this implementation does not (yet) take into account the turning phase - considered negligable compared to distance between waypoints
+// Note that max airspeed available to fight wind is considered currently equal to currently achieved airspeed (by using calc_gndspee_undershoot); should be refined later
+float Location::get_ETA(const struct Location &loc2) const
+{
+	float airspeed_measured;
+	float c_ETA=-1; //Assume first that location is unreachable; use -1 to indicate infinitely long flying due to headwind>=max airspeed; update if location is reachable
+    if (AP::ahrs().airspeed_estimate(airspeed_measured)) {
```

Comment:
I also think you should be using the target airspeed not true, otherwise you cant make any reliable comparisons, the true airspeed will be different each time you ask.

Reply:
...In other words, if we wanted to use the Target Airspeed (instead of measured airspeed), we have to be kinda sure that the User has set that parameter appropriately. Because of the User has - for any reason - set the Target Airspeed incorrectly (ie. not reachable by the Vehicle with its current configuration), then indeed, Measured Airspeed is a safer bet to use. It would be even better to introduce a "max_measured_airspeed" variable, which is always updated to contain the maximum measured or synthetic Airspeed ever reached by the Vehicle any time during the current Mission (in Windy conditions, presumably already having using Throttle correction > cruise Throttle as suggested by calc_gndspeed_undershoot) to check what is _Really_ the maximum available airspeed of the Vehicle. This would be a much safer choice then the user's setting (Presumably). If we over-estimate the Airspeed achiveable at all, then all these calculations (and the proposed original Scrip's calculations ) can lead to a Fly-away, as the Vehicle can not (in reality) achieve the presumably available max Airspeed.

Reply:
If the vehicle is not setup correctly and cannot reach its target airspeed I think that is a issue of tuning that should be fixed before advanced features such as this are used. 

If we did want to try and detect not being able to reach the target airspeed, due to a motor failure or something, that should be its own function that can work all the time rather than just in auto modes.